typescript/infra/scripts/helloworld/kathy.ts

```diff
@@ -54,51 +54,79 @@ async function main() {
+    process.exit(1);
+  }
 
-  // submit frequently so we don't have to wait a super long time for info to get into the metrics
-  const metricsInterval = setInterval(() => {
-    submitMetrics(metricsRegister, 'kathy', { appendMode: true }).catch(
-      console.error,
-    );
-  }, 1000 * 30);
-
-  for (const origin of origins) {
-    for (const destination of origins.filter((d) => d !== origin)) {
-      const labels = {
+  // track how many we are still allowed to send in case some messages send slower than expected.
+  let allowedToSend = 0;
+  setInterval(() => {
+    allowedToSend++;
+  }, fullCycleTime / pairings.length);
+
+  for (
+    // in case we are restarting kathy, keep it from always running the exact same messages first
+    let currentPairingIndex = Date.now() % pairings.length;
+    ;
+    currentPairingIndex = (currentPairingIndex + 1) % pairings.length
+  ) {
+    currentPairingIndexGauge.set(currentPairingIndex);
+    // wait until we are allowed to send the message
+    while (allowedToSend <= 0) await sleep(1000);
+    allowedToSend--;
+
+    const { origin, destination } = pairings[currentPairingIndex];
+    const labels = {
+      origin,
+      remote: destination,
+    };
+    try {
+      await sendMessage(app, origin, destination, gasCalc);
+      log('Message sent successfully', { origin, destination });
+      messagesSendCount.labels({ ...labels, status: 'success' }).inc();
+    } catch (e) {
+      error(`Error sending message, continuing...`, {
+        error: e,
         origin,
-        remote: destination,
-        ...constMetricLabels,
-      };
-      try {
-        await sendMessage(app, origin, destination, gasCalc);
-        messagesSendStatus.labels({ ...labels }).set(1);
-      } catch (err) {
-        console.error(
-          `Error sending message from ${origin} to ${destination}, continuing...`,
-          `${err}`.replaceAll('\n', ' ## '),
-        );
-        failureOccurred = true;
-        messagesSendStatus.labels({ ...labels }).set(0);
-      }
-
-      // Sleep 500ms to avoid race conditions where nonces are reused
-      await sleep(500);
+        destination,
+      });
+      messagesSendCount.labels({ ...labels, status: 'failure' }).inc();
     }
-  }
-
-  clearInterval(metricsInterval);
-  // do not use append mode here so we can clear any old pairings we no longer care about.
-  await submitMetrics(metricsRegister, 'kathy', { appendMode: false });
 
-  if (failureOccurred) {
-    console.error('Failure occurred at least once');
-    process.exit(1);
+    // print stats once every cycle through the pairings
+    if (currentPairingIndex == 0) {
+      for (const [origin, destinationStats] of Object.entries(
+        await app.stats(),
+      )) {
+        for (const [destination, counts] of Object.entries(destinationStats)) {
+          debug('Message stats', { origin, destination, ...counts });
+        }
+      }
+    }
   }
 }
 
+// 10 min by default
+const MESSAGE_RECEIPT_TIMEOUT =
```

Comment:
ditto here with the ms comment. Also may be nice to put this and the configurable full cycle time close to each other in the code so that it's very clear which variables can be configured and how